rllib/connectors/action.py

```diff
@@ -0,0 +1,173 @@
+
+
+register_connector(ConvertToNumpyConnector.__name__, ConvertToNumpyConnector)
+
+
+@DeveloperAPI
+class UnbatchActionsConnector(ActionConnector):
+    """Split action-component batches into single action rows."""
+
+    def __call__(self, ac_data: ActionConnectorDataType) -> ActionConnectorDataType:
+        assert isinstance(
+            ac_data.output, tuple
+        ), "Action connector requires PolicyOutputType data."
+
+        actions, states, fetches = ac_data.output
+        return ActionConnectorDataType(
+            ac_data.env_id, ac_data.agent_id, (unbatch(actions), states, fetches)
+        )
+
+    def to_config(self):
+        return UnbatchActionsConnector.__name__, None
+
+    @staticmethod
+    def from_config(ctx: ConnectorContext, params: List[Any]):
+        return UnbatchActionsConnector(ctx)
+
+
+register_connector(UnbatchActionsConnector.__name__, UnbatchActionsConnector)
+
+
+@DeveloperAPI
+class NormalizeActionsConnector(ActionConnector):
+    def __init__(self, ctx: ConnectorContext):
+        super().__init__(ctx)
+
+        self._action_space_struct = get_base_struct_from_space(ctx.action_space)
+
+    def __call__(self, ac_data: ActionConnectorDataType) -> ActionConnectorDataType:
+        assert isinstance(
+            ac_data.output, tuple
+        ), "Action connector requires PolicyOutputType data."
+
+        actions, states, fetches = ac_data.output
+        return ActionConnectorDataType(
+            ac_data.env_id,
+            ac_data.agent_id,
+            (unsquash_action(actions, self._action_space_struct), states, fetches),
+        )
+
+    def to_config(self):
+        return NormalizeActionsConnector.__name__, None
+
+    @staticmethod
+    def from_config(ctx: ConnectorContext, params: List[Any]):
+        return NormalizeActionsConnector(ctx)
+
+
+register_connector(NormalizeActionsConnector.__name__, NormalizeActionsConnector)
+
+
+@DeveloperAPI
+class ClipActionsConnector(ActionConnector):
+    def __init__(self, ctx: ConnectorContext):
+        super().__init__(ctx)
+
+        self._action_space_struct = get_base_struct_from_space(ctx.action_space)
+
+    def __call__(self, ac_data: ActionConnectorDataType) -> ActionConnectorDataType:
+        assert isinstance(
+            ac_data.output, tuple
+        ), "Action connector requires PolicyOutputType data."
+
+        actions, states, fetches = ac_data.output
+        return ActionConnectorDataType(
+            ac_data.env_id,
+            ac_data.agent_id,
+            (clip_action(actions, self._action_space_struct), states, fetches),
+        )
+
+    def to_config(self):
+        return ClipActionsConnector.__name__, None
+
+    @staticmethod
+    def from_config(ctx: ConnectorContext, params: List[Any]):
+        return ClipActionsConnector(ctx)
+
+
+register_connector(ClipActionsConnector.__name__, ClipActionsConnector)
+
+
+@DeveloperAPI
+class ActionConnectorPipeline(ActionConnector, ConnectorPipeline):
```

Comment:
Still not sure we really need those pipelines (at least at this stage).

Reply:
the pipelines have 2 use cases:
1. util functions to easily manipulate the list of connectors. prepend/append/insert/remove.
2. at least for the agent connectors, the API is "single data piece in, multi-data pieces out", so we need a pipeline to chain things.
let me know what you think.